[cxx-interop] Allow Swift to access non-public C++ members

j-hui opened the pull request:
This patch introduces an a C++ class annotation, SWIFT_PRIVATE_FILEID, which will specify where Swift extensions of that class will be allowed to access its non-public members, e.g.:

    class SWIFT_PRIVATE_FILEID("MyModule/MyFile.swift") Foo { ... };

The goal of this feature is to help C++ developers incrementally migrate the implementation of their C++ classes to Swift, without breaking encapsulation and indiscriminately exposing those classes' private and protected fields.

rdar://137764620

----

This patch is a reimplementation of the feature work from #77726, which got too large to be properly tested and reviewed. I first proposed this feature on Swift Forums: https://forums.swift.org/t/feature-proposal-accessing-non-public-c-members-from-swift/76116.

For those reviewing: I'd like feedback one aspect of this feature/patch, which is the name of the annotation. Specifically, I'm worried about it being named "Swift private" because that that could be confused with the existing `swift_private` attribute (which tells ClangImporter _not_ to import a field).

Several other pieces of follow-up/concurrent work are needed (though none block this patch):

- [x] Another patch to make non-public inherited members accessible in Swift. ~~Those are not being imported yet. (This work was already done in #77726 so it should be pretty straightforward to port.)~~ PR: https://github.com/swiftlang/swift/pull/79348
- [ ] The Swift forums pitch (specifically, the [gist](https://gist.github.com/j-hui/0bab5ce36aef26458b523dd6039af54a) it links to) it a bit outdated; I need to update it to reflect the current implementation (specifically, the behavior for weird annotations etc.).
- [ ] Update release notes/Swift documentation to talk about this feature.

j-hui committed:
[cxx-interop] Allow Swift to access non-public C++ members

This patch introduces an a C++ class annotation, SWIFT_PRIVATE_FILEID,
which will specify where Swift extensions of that class will be allowed
to access its non-public members, e.g.:

    class SWIFT_PRIVATE_FILEID("MyModule/MyFile.swift") Foo { ... };

The goal of this feature is to help C++ developers incrementally migrate
the implementation of their C++ classes to Swift, without breaking
encapsulation and indiscriminately exposing those classes' private and
protected fields.

rdar://137764620

j-hui commented:
@swift-ci please smoke test

Xazax-hun commented on test/Interop/Cxx/class/access/Inputs/non-public.h:
```diff
@@ -0,0 +1,60 @@
+#endif
+
+TEST_CLASS
+__attribute__((__swift_attr__("private_fileid:main/blessed.swift"))) MyClass {
```
Would be nice to have an annotated template to make sure this is propagated to the instantiations properly.

j-hui replied:
Addressed in [f41af53](https://github.com/swiftlang/swift/pull/79093/commits/f41af53c18fdd0c263293b3ec943172817867793)

Xazax-hun commented on include/swift/AST/SourceFile.h:
```diff
@@ -662,6 +662,14 @@ class SourceFile final : public FileUnit {
+
+  /// Parses a file ID string into the module name and file name (with ".swift"
+  /// suffix). Returns nullopt if the file ID string is invalid.
+  static std::optional<std::pair<StringRef, StringRef>>
```
I think if we had a `struct FileIdStr { StringRef module, file; };` that would make most of this comment redundant.  Up to you, I have no strong feelings here.

Xazax-hun replied:
This might give the opportunity to implement `matchesFileID` as an `operator==` on `FileIdStr`.

j-hui replied:
I created a `FileIDStr` struct, but I decided not to use `operator==` since `matches` is more explicit (and still fairly succinct)

Xazax-hun commented on lib/AST/Module.cpp:
```diff
@@ -3649,6 +3649,33 @@ ASTScope &SourceFile::getScope() {
+
+std::optional<std::pair<StringRef, StringRef>>
+SourceFile::parseFileID(StringRef fileID) {
+  auto names = fileID.split('/');
```
Would we always have a single `/`?

j-hui replied:
Yes, I think so. The format of a fileID is always `modulename/filename`.

j-hui replied:
Added a comment to mention this

Xazax-hun commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -8316,3 +8316,19 @@ AccessLevel importer::convertClangAccess(clang::AccessSpecifier access) {
+importer::getPrivateFileIDAttrs(const clang::Decl *decl) {
+  llvm::SmallVector<std::pair<StringRef, clang::SourceLocation>, 1> files;
+
+  auto prefix = StringRef("private_fileid:");
```
Could be constexpr?

Xazax-hun reviewed (approved):
Looks good!

DougGregor commented on lib/AST/DeclContext.cpp:
```diff
@@ -1447,16 +1448,28 @@ bool AccessScope::allowsPrivateAccess(const DeclContext *useDC, const DeclContex
+      return false;
+
+    auto blessedFileID = importer::getPrivateFileIDAttrs(clangDecl);
+    if (blessedFileID.empty() || !useSF->matchesFileID(blessedFileID[0].first))
```
The importer is supporting multiple attributes, so we should loop over all of them here.

DougGregor commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -8316,3 +8316,19 @@ AccessLevel importer::convertClangAccess(clang::AccessSpecifier access) {
+
+  auto prefix = StringRef("private_fileid:");
+
+  if (decl->hasAttrs())
```
It's a very small stylistic thing, but could you put some braces for the outermost "then" statement?

DougGregor commented on lib/ClangImporter/SwiftBridging/swift/bridging:
````diff
@@ -191,6 +191,42 @@
+/// ```
+///
+/// Non-public access is still forbidden outside of extensions and outside of
+/// the designated file ID.
````
Thank you for the clear macro API documentation here

DougGregor commented on include/swift/AST/DiagnosticsClangImporter.def:
```diff
@@ -208,6 +208,24 @@ NOTE(iterator_method_unavailable, none, "C++ method '%0' that returns an "
 NOTE(iterator_potentially_unsafe, none, "C++ methods that return iterators "
                                         "are potentially unsafe; try using Swift collection APIs instead", ())
 
+ERROR(private_fileid_attr_repeated, none,
```
This appears unused.

j-hui replied:
Oops, thanks for catching that! I just checked in the code in `ImportDecl.cpp` that actually emits these diagnostics.

j-hui committed:
Address reviewers comments:

-       Make StringRef("private_fileid:") a constexpr
-       Move fileID handling code into SourceFile::FileIDStr class
-       Document what is a valid fileID format
-       Diagnose multiple/invalid SWIFT_PRIVATE_FILEID annotations
-       Handle multiple SWIFT_PRIVATE_FILEID annotations, even though
        it should not be valid (makes it robust against feature changes
        and diagnostic implementation issues)

j-hui committed:
Also add diagnostic for incomplete types

j-hui commented:
@swift-ci  Please test

j-hui commented:
Thanks for the feedback @DougGregor and @Xazax-hun ! I think I addressed most of the points; I'll add the test case that Gabor suggested tomorrow.

j-hui committed:
Add test case to check that SWIFT_PRIVATE_FILEID works as intended with templates

j-hui committed:
Add test case that checks SWIFT_PRIVATE_FILEID diagnostics

j-hui committed:
Add test case for SWIFT_PRIVATE_FILEID on nested C++ classes

j-hui commented:
@swift-ci Please test

j-hui committed:
Merge remote-tracking branch 'origin/main' into feat-SWIFT_PRIVATE_FILEID

j-hui commented:
@swift-ci  Please test

Xazax-hun commented:
@swift-ci Please test

j-hui commented:
@swift-ci please test windows platform

j-hui commented:
@swift-ci please test windows platform

j-hui committed:
Remove named aggregate initializer (requires C++20)

j-hui commented:
@swift-ci please test

j-hui committed:
Skip private-fileid-diagnostics test on Windows

j-hui commented:
@swift-ci please smoke test
